### PR TITLE
Fetch OAuth client data from API on SSR

### PR DESCRIPTION
**Pull request by yurynix**
This PR changes how OAuth client data is fetched on SSR, it aims to get the data always from the API with caching
 
#### Testing instructions
  
1. Run `git checkout add/ssr-fetch-oauth-client-data` and start your server, or open a [live branch](https://calypso.live/?branch=add/ssr-fetch-oauth-client-data)
2. Make sure:
`curl --resolve calypso.localhost:3000:127.0.0.1 'http://calypso.localhost:3000/log-in?client_id=930' | grep "vaultpress"`
works
3. Make sure 
`curl --resolve calypso.localhost:3000:127.0.0.1 'http://calypso.localhost:3000/log-in?client_id=12' | grep "Picplz"`
works   
 4. Notice the subsequent requests take less time:
```
GET /log-in?client_id=12 200 2137.735 ms - 38078
...
GET /log-in?client_id=12 200 1300.737 ms - 37930
```

#### Reviews
  
- [x] Code
- [x] Product
- [x] Tests

### CODE REVIEW

File: client/login/controller.js [hunk 1]
```diff
@@ -11,32 +11,40 @@ import MagicLogin from './magic-login';
 import HandleEmailedLinkForm from './magic-login/handle-emailed-link-form';
 import { fetchOAuth2ClientData } from 'state/login/oauth2/actions';
 
+const enchanceContextWithWithLogin = context => {
```
Comment (pento) [inline client/login/controller.js]: `enchanceContextWithWithLogin`?  Did you mean `enhanceContextWithLogin`?

**Comment by pento** [comment]
Oh, this is so much better. That heading flicker was killing me. :-)

**Comment by scruffian** [comment]
Product 👍

File: client/state/test/utils.js [hunk 2]
```diff
@@ -682,4 +683,99 @@ describe( 'utils', () => {
...
+				connectedSuccessActionCreator,
+				connectedFailureActionCreator
+			);
+
+			const dispatchResult = actionCreator( 123 )( dispatch );
+			expect( loadingActionCreator ).to.be.calledWith( 123 );
+
+			return dispatchResult.then( () => {
+				expect( successActionCreator ).to.be.calledWith( 'success_data' );
+				//console.log( 'spy', failureActionCreator.getCall( 0 ).args[ 0 ] );
```
Comment (scruffian) [inline client/state/test/utils.js]: Don't forget to remove this

File: client/state/utils.js [hunk 3]
```diff
@@ -409,3 +410,66 @@ export const withoutPersistence = reducer => ( state, action ) => {
...
+ *		dispatch( {
+ *			type: OAUTH2_CLIENT_DATA_REQUEST_FAILURE,
+ *			error,
+ *		} );
+ *
+ *		return Promise.reject( error );
+ *	},
+ *);
+ *
+ * @param {Function} work a worker function that returns the promise ( param1, param2, ... ) => Promise
```
Comment (scruffian) [inline client/state/utils.js]: Do you think `worker` would be a better name?

File: client/state/test/utils.js [hunk 4]
```diff
@@ -682,4 +683,99 @@ describe( 'utils', () => {
...
+
+			return dispatchResult.then( () => {
+				expect( failureActionCreator ).to.be.calledWith( 'error_data' );
+				expect( successActionCreator ).not.to.be.called;
+			} );
+		} );
+
+		it( 'should cache same parameters successful call', () => {
+			const actionCreator = cachingActionCreatorFactory(
+				successfullWork,
```
Comment (scruffian) [inline client/state/test/utils.js]: `successful` is the correct spelling :)

**Comment by scruffian** [comment]
This looks great. Just a few minor comments.

File: client/login/controller.js (hunk 1 above)
Comment (yurynix) [inline client/login/controller.js]: lol, yes =) thanks!

**Comment by pento** [comment]
👍🏻 Let's do this.

**Pull request merged**